pkg/controller/nodeconfig/config/ntp.go

```diff
@@ -0,0 +1,412 @@
+	ctlv1 "github.com/harvester/node-manager/pkg/generated/controllers/node.harvesterhci.io/v1beta1"
+	"github.com/harvester/node-manager/pkg/utils"
+)
+
+const (
+	NTPName                 = "ntp"
+	systemdTimesyncdService = "systemd-timesyncd.service"
+	//etcSystemd                = "/host/etc/systemd/"
+	timesyncdConfigPath       = "/host/etc/systemd/timesyncd.conf"
+	timesyncdConfigBackupPath = "/host/etc/systemd/timesyncd.conf.bak"
+	timesyncdConfigOriginPath = "/host/etc/systemd/timesyncd.conf.origin"
+	timesyncdService          = "systemd-timesyncd"
+	timeWaitSyncService       = "systemd-time-wait-sync"
+	oemPath                   = "/host/oem/"
+	configNTPServer           = "ntpServer"
+)
+
+var etcSystemd = utils.SystemdConfigPath
+
+type NTPStatusAnnotation utils.NTPStatusAnnotation
+
+// AppliedConfigs is a json format string, you should unmarshal it to AppliedConfigAnnotation
+type NTPHandler struct {
+	NTPConfig      *nodeconfigv1.NTPConfig
+	AppliedConfigs string
+	NodeClient     ctlnode.NodeClient
+	ConfName       string
+	mtx            *sync.Mutex
+}
+
+func NewNTPConfigHandler(mtx *sync.Mutex, nodes ctlnode.NodeController, confName string, ntpconfigs *nodeconfigv1.NTPConfig, appliedConfig string) *NTPHandler {
+	return &NTPHandler{
+		NTPConfig:      ntpconfigs,
+		AppliedConfigs: appliedConfig,
+		NodeClient:     nodes,
+		ConfName:       confName,
+		mtx:            mtx,
+	}
+}
+
+// The main thing about update NTP, return bool for restart service and generic error
+func (handler *NTPHandler) DoNTPUpdate(forceUpdate bool) (bool, error) {
+	var content nodeconfigv1.AppliedConfigAnnotation
+	if !forceUpdate && handler.AppliedConfigs != "" {
+		logrus.Infof("Found applied config from annotation: %s", handler.AppliedConfigs)
+		err := json.Unmarshal([]byte(handler.AppliedConfigs), &content)
+		if err != nil {
+			logrus.Warnf("Unmarshal applied config from annotation failed, assume that is empty err: %v", err)
+		}
+
+		if content.NTPServers == handler.NTPConfig.NTPServers {
+			return false, nil
+		}
+	}
+
+	// if the incoming NTPServers is empty but we have annotation, we should remove the NTP config
+	if handler.NTPConfig.NTPServers == "" && handler.AppliedConfigs == "" {
+		return false, nil
+	}
+
+	logrus.Infof("Backup current ntp config ...")
+	if err := handler.backupNTPConfig(); err != nil {
+		return false, fmt.Errorf("backup NTP config failed, skip this round. err: %v", err)
+	}
+
+	logrus.Infof("Prepare to update NTP server with: %s", handler.NTPConfig.NTPServers)
+	if err := handler.updateNTPConfig(); err != nil {
+		return false, fmt.Errorf("update NTP config failed, skip this round. err: %v", err)
+	}
+
+	return true, nil
+}
+
+// we would write the tempfile first then rename to the target.
+func (handler *NTPHandler) updateNTPConfig() error {
+	// the temp config file name will be with suffix .ini
+	tempFileName := etcSystemd + "timesyncd.conf." + "ini"
```

Comment:
Can we remove the `etcSystemd` variable and just use `utils.SystemdConfigPath`.

Reply:
updated!